### PR TITLE
new lint: using &Ref patterns instead of matching on *expr (fixes #187)

### LINKED ISSUE #187: Suggest `match *x { a => 0, b => 1 }` instead of `match x { &a => 0, &b => 1 }`
What the title says. Instead of matching on the borrowed value and then prepending `&` to each match arm, you could also match on the deref'd value. (That's what it's called, right? It's Monday morning.)

This doesn't really change the semantics but shortens the code and makes it a bit clearer (IMHO). I'm not sure if it's really something _clippy_ wants to do (or rather something _rustfmt_ might offer), but I wanted to suggest it anyway.

Example:

```rust
enum Lit {Lorem, Ipsum, Dolor}

fn stuff(lit: &Lit) {
    match lit {
        &Lit::Lorem => println!("1"),
        &Lit::Ipsum => println!("2"),
        &Lit::Dolor => println!("3"),
    }
}

fn main() {
    stuff(&Lit::Lorem);
}
```

([Playpen](http://is.gd/nG6LB3)) can be written as

```rust
enum Lit {Lorem, Ipsum, Dolor}

fn stuff(lit: &Lit) {
    match *lit {
        Lit::Lorem => println!("1"),
        Lit::Ipsum => println!("2"),
        Lit::Dolor => println!("3"),
    }
}

fn main() {
    stuff(&Lit::Lorem);
}
```

([Playpen](http://is.gd/L6QSn1))

This of course only works when all match arms destructure the borrowed value anyway. This seems common enough, though. I just saw [this code](https://github.com/Manishearth/rust-clippy/blob/759b45a46d299f19a10477cc226e70002ac91340/src/consts.rs#L129-L142) in Manishearth/rust-clippy#184.

### PR DESCRIPTION
**Pull request by birkenfeld**
Also move the existing lint that has to do with `match` into the new module.

### CODE REVIEW

File: src/matches.rs [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+fn is_unit_expr(expr: &Expr) -> bool {
+    match expr.node {
+        ExprTup(ref v) if v.is_empty() => true,
+        ExprBlock(ref b) if b.stmts.is_empty() && b.expr.is_none() => true,
+        _ => false,
+    }
+}
+
+fn has_only_ref_pats(arms: &[Arm]) -> bool {
+    for arm in arms {
```
Comment (Manishearth) [inline src/matches.rs]: `arms.iter().flat_map(|a| a.pats).all(|p| match ... )`?

(Iterator gymnastics are fun!)

**Comment by Manishearth** [comment]
lint LGTM, fix the iterator if you want (not necessary)

File: src/matches.rs (hunk 1 above)
Comment (birkenfeld) [inline src/matches.rs]: Sure!

**Comment by llogiq** [comment]
I was a bit surprised that eq_op didn't come up, then again it usually matches on *tuples* of references. I believe this tuple structure actually forbids using the values directly, because to build that, we'd need to move the values.

**Comment by birkenfeld** [comment]
Thanks for merging!

**Comment by llogiq** [comment]
You're welcome.

**Comment by oli-obk** [comment]
lol what is coveralls doing??

**Comment by oli-obk** [comment]
oooh, @birkenfeld did you turn on coveralls in your private repo?

**Comment by birkenfeld** [comment]
I don't think so? The page behind the badge also looks strange, with `.R` files listed as having coverage...

**Pull request merged**